File: LXUtilities/UIKitExtensions/UITableView+LXExtension.m
//
//  UITableView+LXExtension.m
//
//  Created by 从今以后 on 16/6/21.
//  Copyright © 2016年 从今以后. All rights reserved.
//

#import "UIView+LXExtension.h"
#import "UITableView+LXExtension.h"

@implementation UITableView (LXExtension)

- (void)lx_reloadDataWithCompletion:(void (^)(void))completion
{
    [CATransaction begin];
    [CATransaction setCompletionBlock:^{
        completion();
    }];
    [self reloadData];
    [CATransaction commit];
}

- (UITableViewCell *)lx_cellForSelectedRow {
    return [self cellForRowAtIndexPath:[self indexPathForSelectedRow]];
}

- (NSArray<UITableViewCell *> *)lx_cellsForSelectedRows
{
    NSMutableArray *cells = [NSMutableArray new];
    for (NSIndexPath *indexPath in [self indexPathsForSelectedRows]) {
        [cells addObject:[self cellForRowAtIndexPath:indexPath]];
    }
    return [cells copy];
}

- (NSArray<NSIndexPath *> *)lx_allIndexPaths
{
    NSInteger countOfIndexPaths = 0;
    NSInteger countOfSections = [self numberOfSections];
    for (NSInteger section = 0; section < countOfSections; ++section) {
       countOfIndexPaths += [self numberOfRowsInSection:section];
    }
    NSMutableArray *indexPaths = [NSMutableArray arrayWithCapacity:countOfIndexPaths];
    for (NSInteger section = 0; section < countOfSections; ++section) {
        NSInteger countOfRows = [self numberOfRowsInSection:section];
        for (NSInteger row = 0; row < countOfRows; ++row) {
            [indexPaths addObject:[NSIndexPath indexPathForRow:row inSection:section]];
        }
    }
    return [indexPaths copy];
}

- (NSArray<NSIndexPath *> *)lx_indexPathsInSection:(NSInteger)section
{
    NSInteger countOfRows = [self numberOfRowsInSection:section];
    NSMutableArray *indexPaths = [NSMutableArray arrayWithCapacity:countOfRows];
    for (NSInteger row = 0; row < countOfRows; ++row) {
        [indexPaths addObject:[NSIndexPath indexPathForRow:row inSection:section]];
    }
    return [indexPaths copy];
}

- (void)lx_selectRowsAtIndexPaths:(NSArray<NSIndexPath *> *)indexPaths animated:(BOOL)animated
{
    [self beginUpdates];
    for (NSIndexPath *indexPath in indexPaths) {
        [self selectRowAtIndexPath:indexPath animated:animated scrollPosition:UITableViewScrollPositionNone];
    }
    [self endUpdates];
}

void _lx_updateTableHeaderOrFooterViewHeight(UITableView *tableView, BOOL isHeader, void (^configuration)(void))
{
	UIView *headerOrFooterView = isHeader ? tableView.tableHeaderView : tableView.tableFooterView;

	// 添加宽度约束来从而构成完整的约束来计算视图高度
	headerOrFooterView.translatesAutoresizingMaskIntoConstraints = NO;
	NSLayoutConstraint *widthConstraint =
	[NSLayoutConstraint constraintWithItem:headerOrFooterView
								 attribute:NSLayoutAttributeWidth
								 relatedBy:NSLayoutRelationEqual
									toItem:nil
								 attribute:0
								multiplier:1
								  constant:tableView.lx_width];
	[headerOrFooterView addConstraint:widthConstraint];

	// 让调用方有机会更新子视图约束
	!configuration ?: configuration();

	// 通过自动布局计算视图高度，然后移除刚添加的约束
	CGFloat height = [headerOrFooterView systemLayoutSizeFittingSize:UILayoutFittingCompressedSize].height;
	[headerOrFooterView removeConstraint:widthConstraint];

	// 回归传统 frame 布局后设置高度
	headerOrFooterView.translatesAutoresizingMaskIntoConstraints = YES;
	headerOrFooterView.lx_height = height;

	if (isHeader) {
		tableView.tableHeaderView = headerOrFooterView;
	} else {
		tableView.tableFooterView = headerOrFooterView;
	}
}

- (void)lx_updateTableHeaderViewHeightWithLayoutConfiguration:(void (^)(void))configuration {
	_lx_updateTableHeaderOrFooterViewHeight(self, YES, configuration);
}

- (void)lx_updateTableFooterViewHeightWithLayoutConfiguration:(void (^)(void))configuration {
    _lx_updateTableHeaderOrFooterViewHeight(self, NO, configuration);
}

@end

File: LXUtilities/UIKitExtensions/Swift/UITableView+LXExtension.swift
//
//  UITableView+LXExtension.swift
//
//  Created by 从今以后 on 2017/9/19.
//  Copyright © 2017年 从今以后. All rights reserved.
//

import UIKit

// MARK: - 数据源
extension Swifty where Base: UITableView {

    func registerReusableCell<T: UITableViewCell & ReusableView>(_: T.Type) {
		base.register(T.self, forCellReuseIdentifier: T.reuseIdentifier)
    }

	func registerReusableCell<T: UITableViewCell & ReusableView & NibLoadableView>(_: T.Type) {
		base.register(T.nib, forCellReuseIdentifier: T.reuseIdentifier)
	}

    // e.g. tableView.dequeueReusableCell(for: indexPath) as CustomCell
    func dequeueReusableCell<T: UITableViewCell & ReusableView>(for indexPath: IndexPath) -> T {
        return base.dequeueReusableCell(withIdentifier: T.reuseIdentifier, for: indexPath) as! T
    }

    func registerReusableHeaderFooterView<T: UITableViewHeaderFooterView & ReusableView>(_: T.Type) {
		base.register(T.self, forHeaderFooterViewReuseIdentifier: T.reuseIdentifier)
    }

	func registerReusableHeaderFooterView<T: UITableViewHeaderFooterView & ReusableView & NibLoadableView>(_: T.Type) {
		base.register(T.nib, forHeaderFooterViewReuseIdentifier: T.reuseIdentifier)
	}

    func dequeueReusableHeaderFooterView<T: UITableViewHeaderFooterView & ReusableView>() -> T {
        return base.dequeueReusableHeaderFooterView(withIdentifier: T.reuseIdentifier) as! T
    }
}

// MARK: - 刷新数据
extension Swifty where Base: UITableView {

    func reloadData(completion: @escaping () -> Void) {
        CATransaction.begin()
        CATransaction.setCompletionBlock {
            completion()
        }
        base.reloadData()
        CATransaction.commit()
    }
}

// MARK: - 索引
extension Swifty where Base: UITableView {

    func indexPathsInSection(_ section: Int) -> [IndexPath]? {
        let indexPaths = (0..<base.numberOfRows(inSection: section)).map {
            IndexPath(row: $0, section: section)
        }
        return indexPaths.isEmpty ? nil : indexPaths
    }

    func allIndexPaths() -> [IndexPath]? {
        let indexPaths = (0..<base.numberOfSections).flatMap { (section) -> [IndexPath] in
            (0..<base.numberOfRows(inSection: section)).map { (row) in
                IndexPath(row: row, section: section)
            }
        }
        return indexPaths.isEmpty ? nil : indexPaths
    }
    
    func indexPathsForSelectedRowsInSection(_ section: Int) -> [IndexPath]? {
        return base.indexPathsForSelectedRows?.filter { $0.section == section }
    }
}

// MARK: - 单元格
extension Swifty where Base: UITableView {

    func cellForSelectedRow() -> UITableViewCell? {
        guard let indexPath = base.indexPathForSelectedRow else {
            return nil
        }
        guard let cell = base.cellForRow(at: indexPath) else {
            return nil
        }
        return cell
    }

    func cellsForSelectedRows() -> [UITableViewCell]? {
        return base.indexPathsForSelectedRows?.compactMap { base.cellForRow(at: $0) }
    }
}

// MARK: - 选中
extension Swifty where Base: UITableView {

    func selectRows(at indexPaths: [IndexPath], animated: Bool) {
        UIView.performWithoutAnimation {
            base.beginUpdates()
            indexPaths.forEach {
                base.selectRow(at: $0, animated: animated, scrollPosition: .none)
            }
            base.endUpdates()
        }
    }

    func selectRows(inSection section: Int, animated: Bool) {
        guard let indexPaths = indexPathsInSection(section) else {
            return
        }
        selectRows(at: indexPaths, animated: animated)
    }

    func selectAllRows() {
        guard let indexPaths = allIndexPaths() else {
            return
        }
        UIView.performWithoutAnimation {
            base.beginUpdates()
            indexPaths.forEach {
                base.selectRow(at: $0, animated: false, scrollPosition: .none)
            }
            base.endUpdates()
        }
    }
}

// MARK: - 反选
extension Swifty where Base: UITableView {

    func deselectRows(at indexPaths: [IndexPath], animated: Bool) {
        UIView.performWithoutAnimation {
            base.beginUpdates()
            indexPaths.forEach {
                base.deselectRow(at: $0, animated: animated)
            }
            base.endUpdates()
        }
    }

    func deselectRows(inSection section: Int, animated: Bool) {
        guard let indexPaths = indexPathsInSection(section) else {
            return
        }
        deselectRows(at: indexPaths, animated: animated)
    }

    func deselectAllRows() {
        guard let indexPaths = base.indexPathsForSelectedRows else {
            return
        }
        UIView.performWithoutAnimation {
            base.beginUpdates()
            indexPaths.forEach {
                base.deselectRow(at: $0, animated: false)
            }
            base.endUpdates()
        }
    }
}

// MARK: - 头尾视图
extension Swifty where Base: UITableView {

    func updateTableHeaderViewHeight(withLayoutConfiguration configuration: (() -> Void)? = nil) {
        let headerView = base.tableHeaderView!
        headerView.translatesAutoresizingMaskIntoConstraints = false
        let widthConstraint = headerView.widthAnchor.constraint(equalToConstant: base.lx.width)
        widthConstraint.isActive = true
        configuration?()
        let height = headerView.systemLayoutSizeFitting(UIView.layoutFittingCompressedSize).height
        widthConstraint.isActive = false
        headerView.translatesAutoresizingMaskIntoConstraints = true
        headerView.lx.height = height
        base.tableHeaderView = headerView
    }

    func updateTableFooterViewHeight(withLayoutConfiguration configuration: (() -> Void)? = nil) {
        let footerView = base.tableFooterView!
        footerView.translatesAutoresizingMaskIntoConstraints = false
        let widthConstraint = footerView.widthAnchor.constraint(equalToConstant: base.lx.width)
        widthConstraint.isActive = true
        configuration?()
        let height = footerView.systemLayoutSizeFitting(UIView.layoutFittingCompressedSize).height
        widthConstraint.isActive = false
        footerView.translatesAutoresizingMaskIntoConstraints = true
        footerView.lx.height = height
        base.tableFooterView = footerView
    }
}
